### PR TITLE
Switch to lucene default reader

### PR DESCRIPTION
**Pull request by sam-herman**
### Description
Switch jVector reader to use the current Lucene directory implementation which is compatible with JDK21.
- Fixes the 2GB segment limit while eliminate the need to force JDK21+
- Update documentation

### Check List
- [x ] New functionality includes testing.
~- [ ] New functionality has been documented.~
~- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md).~
- [x ] Commits are signed per the DCO using `--signoff`.
~- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose).~

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/k-NN/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/main/java/org/opensearch/knn/index/codec/jvector/JVectorReader.java [hunk 1]
```diff
@@ -151,6 +139,7 @@ public void search(String field, byte[] target, KnnCollector knnCollector, Bits
     public void close() throws IOException {
         for (FieldEntry fieldEntry : fieldEntryMap.values()) {
             IOUtils.close(fieldEntry.readerSupplier::close);
+            // IOUtils.close(fieldEntry.index::close);
```
Comment (tjake) [inline src/main/java/org/opensearch/knn/index/codec/jvector/JVectorReader.java:142]: remove?

File: src/main/java/org/opensearch/knn/index/codec/jvector/JVectorRandomAccessReader.java [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+    }
+
+    @Override
+    public void readFully(byte[] bytes) throws IOException {
+        indexInputDelegate.readBytes(bytes, 0, bytes.length);
+    }
+
+    @Override
+    public void readFully(ByteBuffer buffer) throws IOException {
+        indexInputDelegate.readBytes(buffer.array(), buffer.arrayOffset(), buffer.remaining());
```
Comment (tjake) [inline src/main/java/org/opensearch/knn/index/codec/jvector/JVectorRandomAccessReader.java:67]: this should be:

indexInputDelegate.readBytes(buffer.array(), buffer.arrayOffset() + buffer.position(), buffer.remaining());

File: src/main/java/org/opensearch/knn/index/codec/jvector/JVectorRandomAccessReader.java [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+    }
+
+    @Override
+    public void readFully(long[] vector) throws IOException {
+        for (int i = 0; i < vector.length; i++) {
```
Comment (tjake) [inline src/main/java/org/opensearch/knn/index/codec/jvector/JVectorRandomAccessReader.java:72]: these may be faster if you get the bytes into a BB then extract the longs.  Same with all the others.

**Review by tjake — CHANGES_REQUESTED** [review]

File: src/main/java/org/opensearch/knn/index/codec/jvector/JVectorRandomAccessReader.java (hunk 2 above)
Comment (sam-herman) [reply src/main/java/org/opensearch/knn/index/codec/jvector/JVectorRandomAccessReader.java:67]: You prompted me to now actually notice that the right option should have been:
```java
        while (buffer.hasRemaining() && indexInputDelegate.getFilePointer() < indexInputDelegate.length()) {
            indexInputDelegate.readBytes(this.buffer, 0, Long.BYTES);
            buffer.put(this.buffer, 0, Long.BYTES);
        }
```
As I previously omitted the use of the local buffer and this way we don't have to deal with any edge condition of the BB implementation.

Comment (sam-herman) [reply src/main/java/org/opensearch/knn/index/codec/jvector/JVectorRandomAccessReader.java:67]: As discussed offline, need to add more conditions here and tests to make sure we don't fall into those edge cases of `buffer.remaining() < whateverReaminsOnStream`

File: src/main/java/org/opensearch/knn/index/codec/jvector/JVectorReader.java (hunk 1 above)
Comment (sam-herman) [reply src/main/java/org/opensearch/knn/index/codec/jvector/JVectorReader.java]: done

File: src/main/java/org/opensearch/knn/index/codec/jvector/JVectorRandomAccessReader.java (hunk 3 above)
Comment (tjake) [reply src/main/java/org/opensearch/knn/index/codec/jvector/JVectorRandomAccessReader.java:93]: all good
